build_tools/python/e2e_test_framework/serialization_test.py

```diff
@@ -0,0 +1,142 @@
+## Copyright 2022 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from dataclasses import dataclass
+import json
+import pathlib
+from typing import Dict, List, Optional
+import collections
+import enum
+import typing
+import unittest
+
+from e2e_test_framework import serialization
+
+
+class EnumX(enum.Enum):
+  OPTION_A = "a"
+  OPTION_B = "b"
+  OPTION_C = "c"
+
+
+@serialization.serializable
+@dataclass
+class TestC(object):
+  float_val: float
+
+
+@serialization.serializable(keyed_obj=True, type_key="test_b", id_field="key")
+@dataclass
+class TestB(object):
+  key: str
+  int_val: int
+
+
+@serialization.serializable
+@dataclass
+class TestA(object):
+  b_list: List[TestB]
+  c_obj: TestC
+  str_val: Optional[str]
+  enum_val: EnumX
+
+
+@serialization.serializable
+@dataclass
+class TestUnsupported(object):
+  path: pathlib.PurePath
+
+
+@serialization.serializable(keyed_obj=True, type_key="test_circular")
+@dataclass
+class TestCircularReference(object):
+  id: str
+  child: Optional["TestCircularReference"]
+
+
+class SerializationTest(unittest.TestCase):
+  B_OBJ_A = TestB(key="id_a", int_val=10)
+  B_OBJ_B = TestB(key="id_b", int_val=20)
+  TEST_OBJS = [
+      TestA(b_list=[B_OBJ_A, B_OBJ_B],
+            c_obj=TestC(float_val=0.1),
+            str_val="test1",
+            enum_val=EnumX.OPTION_B),
+      TestA(b_list=[B_OBJ_A],
+            c_obj=TestC(float_val=0.2),
+            str_val=None,
+            enum_val=EnumX.OPTION_C)
+  ]
```

Comment:
I'd recommend against moving these into class fields. Each test function should stand on its own and this hides relevant properties of the test

Reply:
Done.